lib/charms/opensearch/v0/opensearch_plugins.py

```diff
@@ -0,0 +1,277 @@
+                if remove_keys:
+                    self.distro.remove_from_keystore(key)
+                else:
+                    self.distro.add_to_keystore(key, value, force=force)
+            # Now, reload the security settings and return if opensearch needs restart
+            post = self._request("POST", "_nodes/reload_secure_settings")
+            logger.debug(f"_update_keystore_and_reload: response received {post}")
+        except OpenSearchKeystoreError as ek:
+            raise ek
+        except Exception as e:
+            logger.exception(e)
+            raise OpenSearchPluginError("Unknown error during keystore reload")
+        if post["status"] < 200 or post["status"] >= 300:
+            raise OpenSearchPluginError("Error while processing _nodes reload")
+
+    @property
+    def name(self) -> str:
+        """Returns the name of the plugin."""
+        return self._name
+
+    def uninstall(
+        self,
+        opensearch_yml: Dict[str, str],
+        keystore: Dict[str, str],
+    ) -> bool:
+        """Erases relevant data for this plugin. Returns True if restart needed."""
+        if not self._is_started():
+            return False
+        self._update_keystore_and_reload(keystore)
+        return any(
+            [
+                self.distro.config.delete(self.CONFIG_YML, c, v)[c] == v
+                for c, v in opensearch_yml.items()
+            ]
+        )
+
+    @property
+    def needs_upgrade(self) -> bool:
+        """Returns if the plugin needs an upgrade or not.
+
+        Needs upgrade must be set if an upgrade on the charm happens and plugins must
+        be updated.
+
+        Consider overriding this method only if the plugin needs an special care at upgrade.
+        """
+        current_version = self.version
+        # Now check if the format of this plugin and OpenSearch's match
+        num_points = len(self.distro.version.split("."))
+        return self.distro.version != current_version[:num_points]
+
+    @abstractmethod
+    def upgrade(self, uri: str) -> None:
+        """Runs the upgrade process in this plugin."""
+        pass
+
+    @abstractmethod
+    def is_enabled(self) -> bool:
+        """Returns True if the plugin is enabled."""
+        pass
+
+    @abstractmethod
+    def disable(self) -> bool:
+        """Disables the plugin.
+
+        Runs the configure() method with the configuration to disable the plugin.
+        """
+        pass
+
+    @abstractmethod
+    def enable(self) -> bool:
+        """Enables the plugin.
+
+        Runs the configure() method with the configuration to enable the plugin.
+        """
+        pass
+
+    def configure(
+        self,
+        opensearch_yml: Dict[str, str],
+        keystore: Dict[str, str] = {},
+    ) -> bool:
+        """Sets the plugin configuration. Returns True if restart needed."""
+        if not self._is_started():
+            return False
+        self._update_keystore_and_reload(keystore)
```

Comment:
why is the keystore such a central piece of plugins management?

Reply:
@Mehdi-Bendriss overall, you must add a plugin configuration to opensearch.yml or the default keystore (if that given option is a  ```SecureSetting``` class).
In the case of backup, for example, we can see that access- and secret-keys must be set in keystore, as [they are secure settings](https://github.com/opensearch-project/OpenSearch/blob/bb7d23ca3c31140b03899d83648948dbe7229cf2/plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3ClientSettings.java#L100-L118)

Also, if we decide to remove a plugin but do not clean the keystore, then opensearch fails to start.